zerver/migrations/0147_realm_emoji_drop_unique_constraint.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+import os
+import shutil
+import tempfile
+
+from typing import Dict, Optional
+
+from boto.s3.bucket import Bucket
+from boto.s3.connection import S3Connection
+from boto.s3.key import Key
+from django.conf import settings
+from django.db import migrations, models
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+class Uploader:
+    def __init__(self) -> None:
+        self.old_orig_image_path_template = "{realm_id}/emoji/{emoji_file_name}.original"
+        self.old_path_template = "{realm_id}/emoji/{emoji_file_name}"
+        self.new_orig_image_path_template = "{realm_id}/emoji/images/{emoji_file_name}.original"
+        self.new_path_template = "{realm_id}/emoji/images/{emoji_file_name}"
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        raise NotImplementedError()
+
+    def ensure_emoji_images(self, realm_id: int, old_filename: str, new_filename: str) -> None:
+        # Copy original image file.
+        old_file_path = self.old_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=old_filename)
+        new_file_path = self.new_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+        # Copy resized image file.
+        old_file_path = self.old_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=old_filename)
+        new_file_path = self.new_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+class LocalUploader(Uploader):
+    def __init__(self) -> None:
+        super().__init__()
+
+    @staticmethod
+    def mkdirs(path: str) -> None:
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname):
+            os.makedirs(dirname)
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        src_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', src_path)
+        self.mkdirs(src_path)
+        dst_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', dst_path)
+        self.mkdirs(dst_path)
+        shutil.copyfile(src_path, dst_path)
```

Comment:
Maybe should do copy file only if dst_path already doesn't exist.

Reply:
Don't we know by construction that it doesn't?

Reply:
I was thinking about cases when migrations get interrupted.

Reply:
Ahh, I think in that case we just want to overwrite.  We should be sure this does this.

Reply:
yeah `shutil.copyfile` does overwrite.